gtk/gtkfilechooserwidget.c

```diff
@@ -6180,15 +6180,32 @@ gtk_file_chooser_widget_get_default_size (GtkFileChooserEmbed *chooser_embed,
   GtkFileChooserWidget *impl = GTK_FILE_CHOOSER_WIDGET (chooser_embed);
   GtkFileChooserWidgetPrivate *priv = impl->priv;
   GtkRequisition req;
-  int x, y, width, height;
+  int x, y, width, height, max_width, max_height;
   GSettings *settings;
+  GdkScreen *screen;
+  GdkWindow *window;
+  gint monitor_num;
+  GdkRectangle monitor;
+
+  screen = gtk_widget_get_screen (GTK_WIDGET (impl));
+  window = gtk_widget_get_window (GTK_WIDGET (impl));
+  if (screen && window)
+    {
+      monitor_num = gdk_screen_get_monitor_at_window (screen, window);
+      gdk_screen_get_monitor_workarea (screen, monitor_num, &monitor);
+      max_width = monitor.width * 5 / 6;
+      max_height = monitor.height * 5 / 6;
+    }
+  else
+    max_width = max_height = G_MAXINT;
```

Comment:
nit: you could move this block to a function _gtk_file_chooser_get_monitor_workarea_for_widget()